# Phylogenetic network

A phylogenetic network was produced using the SNP data concatenated for all autosomes, with all samples including the Marsh Wren.

The `.nex` data matrix used for SplitsTree is provided in the folder `4_SplitsTree`. This matrix contains genotypes (for which 0 and 2 are homozygous states for the two alleles at each locus, 1 is heterozygous, and ? is missing data) It can be opened in SplitsTree, which should automatically produce a phylogenetic network within a few seconds. The sample #64 labelled "MAWR" is the outgroup, which can be used to root the phylogenetic network. To set this sample as the outgroup, do: `Edit->Select Nodes-> click on the MAWR node of the tree -> Data -> Set Outgroup`. Then, a rooted equal angle tree can be produced: `Apply Rooted EqualAngle Method -> Apply`. 

SplitsTree also produces a distance matrix for all samples; this matrix is also provided in the `4_SplitsTree` folder. This matrix was used to compare the average distances within and amongst the taxa. Samples are placed in rows and columns in the same order as the `.nex` file. I used this data matrix to calculate the mean distances between and within each species (as well as their standard deviations). I divided the resulting distances by the mean distances amongst Pacific Wrens in order to standardize the data. 